src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs

```diff
@@ -0,0 +1,353 @@
+        {
+            uint mask = TNegator.ExtractMask(result);
+            int offsetInVector = BitOperations.TrailingZeroCount(mask);
+            if (offsetInVector >= Vector128<short>.Count)
+            {
+                // We matched within the second vector
+                current0 = ref current1;
+                offsetInVector -= Vector128<short>.Count;
+            }
+            return offsetInVector + (int)(Unsafe.ByteOffset(ref searchSpace, ref current0) / sizeof(short));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int ComputeLastIndex<TNegator>(ref short searchSpace, ref short current, Vector128<byte> result)
+            where TNegator : struct, INegator
+        {
+            uint mask = TNegator.ExtractMask(result) & 0xFFFF;
+            int offsetInVector = 31 - BitOperations.LeadingZeroCount(mask);
+            return offsetInVector + (int)(Unsafe.ByteOffset(ref searchSpace, ref current) / sizeof(char));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int ComputeLastIndexOverlapped<TNegator>(ref short searchSpace, ref short secondVector, Vector128<byte> result)
+            where TNegator : struct, INegator
+        {
+            uint mask = TNegator.ExtractMask(result) & 0xFFFF;
+            int offsetInVector = 31 - BitOperations.LeadingZeroCount(mask);
+            if (offsetInVector < Vector128<short>.Count)
+            {
+                return offsetInVector;
+            }
+
+            // We matched within the second vector
+            return offsetInVector - Vector128<short>.Count + (int)(Unsafe.ByteOffset(ref searchSpace, ref secondVector) / sizeof(char));
+        }
+
+        private interface INegator
+        {
+            static abstract bool NegateIfNeeded(bool result);
+            static abstract Vector128<byte> NegateIfNeeded(Vector128<byte> result);
+            static abstract uint ExtractMask(Vector128<byte> result);
+        }
+
+        private struct DontNegate : INegator
+        {
+            public static bool NegateIfNeeded(bool result) => result;
+            public static Vector128<byte> NegateIfNeeded(Vector128<byte> result) => result;
+            public static uint ExtractMask(Vector128<byte> result) => ~Vector128.Equals(result, Vector128<byte>.Zero).ExtractMostSignificantBits();
+        }
+
+        private struct Negate : INegator
+        {
+            public static bool NegateIfNeeded(bool result) => !result;
+            // This is intentionally testing for equality with 0 instead of "~result".
+            // We want to know if any character didn't match, as that means it should be treated as a match for the -Except method.
+            public static Vector128<byte> NegateIfNeeded(Vector128<byte> result) => Vector128.Equals(result, Vector128<byte>.Zero);
+            public static uint ExtractMask(Vector128<byte> result) => result.ExtractMostSignificantBits();
+        }
```

Comment:
Don't we already have some of these defined in SpanHelpers or MemoryExtensions?  Is it possible to reuse and/or consolidate the types and associated helpers?

Reply:
We have structs named like these, but the ones in this file rely heavily on the implementation details - e.g. `ExtractMask` does more work than strictly extracting the mask in `DontNegate`, or `NegateIfNeeded` does something slightly different than strictly just negating in `Negate`.

The set of helpers in `ProbabilisticMap.cs` can probably delegate to the one in `MemoryExtensions` though.

Reply:
Changed the ones in `ProbabilisticMap` to reuse the ones in `SpanHelpers`.